### PR TITLE
Add a GitHub Actions workflow for deploying to App Engine

### PR DESCRIPTION
**Pull request by LujieZ**
This PR adds a GitHub Actions workflow that deploys to Google App Engine for
each create/update of a pull request and each push to the main branch. The PR
also includes an `app.yaml` App Engine configuration file for the deployment
(the Node.js runtime is used but only the support for static files is used).

The formats of the deployed App Engine version are `pr#-<commit hash prefix>`
for pull requests and `main-<commit hash prefix>` for pushes to the main
branch.

### CODE REVIEW

File: .github/workflows/appengine-deploy.yml [hunk 1]
```diff
@@ -0,0 +1,47 @@
+name: App Engine Deploy Integration
+
+on: 
+  push:
+    branches:
+      - main 
+  pull_request:
```
Comment (zianc) [inline .github/workflows/appengine-deploy.yml:7]: If there's nothing listed under here, is this necessary?

Comment (LujieZ) [reply .github/workflows/appengine-deploy.yml:7]: It means that pull requests open on any branch should be checked! :)

File: app.yaml [hunk 2]
```diff
@@ -0,0 +1,21 @@
+runtime: nodejs10
+ 
+handlers:
+- url: /
+  static_files: web/src/index.html
```
Comment (razvanm) [inline app.yaml:5]: Running `npm run build` in the `web/` dir will create a `dist/` dir. That's the content we want to use.

File: .github/workflows/appengine-deploy.yml [hunk 3]
```diff
@@ -0,0 +1,47 @@
...
+    name: App Engine Deploy Integration
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out
+        uses: actions/checkout@v2
+
+      - id: vars
+        name: Declare sha_short as a variable
+        shell: bash
+        run: echo "::set-output name=sha_short::$(git rev-parse --short=7 ${{ github.sha }})"
```
Comment (razvanm) [inline .github/workflows/appengine-deploy.yml:23]: We use both `set-env` and `set-output`. Is there a reason why need to use both? Any reason to only use `set-env` that manipulates output parameters?

File: .github/workflows/appengine-deploy.yml [hunk 4]
```diff
@@ -0,0 +1,47 @@
...
+jobs:
+  appengine-deploy:
+    name: App Engine Deploy Integration
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out
+        uses: actions/checkout@v2
+
+      - id: vars
+        name: Declare sha_short as a variable
```
Comment (razvanm) [inline .github/workflows/appengine-deploy.yml:21]: Is the _variable_ term correct? Based on [1] it seems to be called _output parameter_.

[1] https://docs.github.com/en/actions/reference/workflow-commands-for-github-actions#setting-an-output-parameter

File: .github/workflows/appengine-deploy.yml [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+
+jobs:
+  appengine-deploy:
+    name: App Engine Deploy Integration
+    runs-on: ubuntu-latest
+    steps:
+      - name: Check out
+        uses: actions/checkout@v2
+
+      - id: vars
```
Comment (razvanm) [inline .github/workflows/appengine-deploy.yml:20]: Is the `id` required? We don't seem to set for every step.

File: .github/workflows/appengine-deploy.yml [hunk 6]
```diff
@@ -0,0 +1,46 @@
...
+
+      - name: Set APPENGINE_VERSION to main
+        if: github.ref == 'refs/heads/main'
+        run: echo '::set-env name=APPENGINE_VERSION::main'
+
+      - name: Set APPENGINE_VERSION to current PR
+        if: github.ref != 'refs/heads/main'
+        run: echo '::set-env name=APPENGINE_VERSION::pr${{ github.event.pull_request.number }}'
+
+        # Use Google Cloud credentails to deploy an App Engine Application:
```
Comment (razvanm) [inline .github/workflows/appengine-deploy.yml:32]:
```suggestion
        # Use Google Cloud credentials to deploy an App Engine Application:
```

File: app.yaml [hunk 7]
```diff
@@ -0,0 +1,27 @@
+runtime: nodejs10
+ 
+handlers:
+- url: /
+  static_files: dist/web/index.html
+  upload: dist/web/index.html
```
Comment (razvanm) [inline app.yaml:6]: For this path to work we need to move this `app.yaml` to the `web/` dir.

File: app.yaml [hunk 8]
```diff
@@ -0,0 +1,27 @@
...
+
+- url: /(.*\.js)
+  redirect_http_response_code: 301
+  static_files: dist/web/\1
+  upload: dist/web.*\.js
```
Comment (razvanm) [inline app.yaml:11]:
```suggestion
  upload: dist/web/.*\.js
```

File: app.yaml [hunk 9]
```diff
@@ -0,0 +1,27 @@
...
+
+- url: /(.*\.js.map)
+  redirect_http_response_code: 301
+  static_files: dist/web/\1
+  upload: dist/web.*\.js.map
```
Comment (razvanm) [inline app.yaml:16]:
```suggestion
  upload: dist/web/.*\.js.map
```

File: app.yaml [hunk 10]
```diff
@@ -0,0 +1,27 @@
...
+- url: /favicon.ico
+  redirect_http_response_code: 301
+  static_files: dist/web/favicon.ico
+  upload: dist/web/favicon.ico
+
+
+- url: /.*
+  redirect_http_response_code: 301
+  static_files: dist/web/index.html
+  upload: dist/web.*
```
Comment (razvanm) [inline app.yaml:27]:
```suggestion
  upload: dist/web/.*
```

File: app.yaml [hunk 11]
```diff
@@ -0,0 +1,27 @@
+runtime: nodejs10
+ 
+handlers:
```
Comment (razvanm) [inline app.yaml:3]: I think we might be able to trim this to something smaller like this:

```
handlers:
- url: /.*
  static_files: dist/web/\1
  upload: dist/web/.*
```

File: .github/workflows/appengine-deploy.yml [hunk 12]
```diff
@@ -0,0 +1,52 @@
+name: App Engine Deploy Integration
+
+on: 
+  push:
+    branches:
+      - main 
+  pull_request:
+
+env:
+  APPENGINE_VERSION: invalid
```
Comment (razvanm) [inline .github/workflows/appengine-deploy.yml:10]: This doesn't seem to be the prefix for the App Engine version so how about calling this `APPENGINE_VERSION_PREFIX`? Another option is to keep the name and add `-${{ env.SHA_SHORT }}`

File: .github/workflows/appengine-deploy.yml [hunk 13]
```diff
@@ -0,0 +1,52 @@
...
+    steps:
+      - name: Check out
+        uses: actions/checkout@v2
+
+      - name: Set SHA_SHORT to short commit sha
+        run: echo "::set-env name=SHA_SHORT::$(git rev-parse --short=7 ${{ github.event.pull_request.head.sha }})"
+
+      - name: Set APPENGINE_VERSION to main
+        if: github.ref == 'refs/heads/main'
+        run: echo '::set-env name=APPENGINE_VERSION::main'
```
Comment (razvanm) [inline .github/workflows/appengine-deploy.yml:26]: Is the `echo` part useful?

File: web/app.yaml [hunk 14]
```diff
@@ -0,0 +1,27 @@
...
+- url: /favicon.ico
+  redirect_http_response_code: 301
+  static_files: dist/web/favicon.ico
+  upload: dist/web/favicon.ico
+
+
+- url: /.*
+  redirect_http_response_code: 301
+  static_files: dist/web/index.html
+  upload: dist/web/index.html
```
Comment (razvanm) [inline web/app.yaml:27]: Do we need these lines? I would expect the handler for `/.*` starting on line 4 to do everything we need.

File: app.yaml (hunk 11 above)
Comment (LujieZ) [reply app.yaml]: The rest of the file is trimmed and the app.yaml file is updated.

File: .github/workflows/appengine-deploy.yml [hunk 15]
```diff
@@ -0,0 +1,52 @@
...
+
+      - name: Build
+        working-directory: web
+        run: |
+          npm ci
+          npm run build
+
+        # Use Google Cloud credentials to deploy an App Engine Application:
+        # https://github.com/GoogleCloudPlatform/github-actions/tree/master/appengine-deploy
+      - id: deploy
```
Comment (razvanm) [inline .github/workflows/appengine-deploy.yml:40]:
```suggestion
      # Use Google Cloud credentials to deploy an App Engine Application:
      # https://github.com/GoogleCloudPlatform/github-actions/tree/master/appengine-deploy
      - id: deploy
```

**Pull request merged**